File: 2024/3/solve2.py
import re

from pathlib import Path
from operator import mul
from functools import reduce

content = Path("input").read_text()

pattern = r"(do\(\)|don't\(\))|mul\((\d{1,3}),(\d{1,3})\)"

findings = re.findall(pattern, content)

do = True

results = []

for finding in findings:
    if finding[0]:
        do = finding[0] == "do()"
    if finding[0] == "" and do:
        results.append(reduce(mul, map(int, finding[1:]), 1))
print(sum(results))

File: 2024/3/solve2.ex
#!/usr/bin/env elixir

content = File.read!("input")

pattern = ~r/(do|don't|mul)\((?:(\d{1,3}),(\d{1,3}))?\)/

defmodule Stuff do
  def do_thing(["do"], %{"acc" => acc}) do
    %{"enabled" => true, "acc" => acc}
  end
  def do_thing(["don't"], %{"acc" => acc}) do
    %{"enabled" => false, "acc" => acc}
  end
  def do_thing(["mul", n1, n2], %{"enabled" => enabled, "acc" => acc}) when enabled do
    %{"enabled" => true, "acc" => acc + String.to_integer(n1) * String.to_integer(n2)}
  end
  def do_thing(["mul", _n1, _n2], state) do
    state
  end
end
 
state = %{"enabled" => true, "acc" => 0}

pattern
  |> Regex.scan(content)
  |> Enum.map(&tl/1)
  |> Enum.reduce(state, fn match, acc -> Stuff.do_thing(match, acc) end)
  |> IO.inspect()